File: src/clu/sources/dnf_checkupdate.py
import logging

from clu.provides import Provides
from clu.requires import Requires
from clu.facts import Facts
from clu.sources import Source
from clu.input import text_program
from clu.sources import PARSE_FAIL_MSG, NET_DISABLED_MSG
from clu.config import get_config

log = logging.getLogger(__name__)
cfg = get_config()


class DnfCheckUpdate(Source):
    def provides(self, provides: Provides) -> None:
        provides["run.update_required"] = self

    def requires(self, requires: Requires) -> None:
        requires.programs.extend(["dnf check-update"])

    def parse(self, facts: Facts) -> None:
        if not cfg.net:
            facts["run.update_required"] = NET_DISABLED_MSG
            return

        _, rc = text_program("dnf check-update")
        log.debug(f"rc is {rc}")

        if rc == 0:
            value = "False"
        elif rc == 100:
            value = "True"
        else:
            value = PARSE_FAIL_MSG

        facts["run.update_required"] = value

File: pkg/facts/sources/dnf_checkupdate.go
package sources

// Determines whether system package updates are required via `dnf check-update`.
// Exit code semantics (per dnf):
//   0   -> no updates available
//   100 -> updates available
//   other -> error / parsing failure

import (
	"github.com/huntermatthews/clu/pkg/facts/types"
	"github.com/huntermatthews/clu/pkg/global"
	"github.com/huntermatthews/clu/pkg/input"
)

// DnfCheckUpdate reports if updates are required (True/False) or an error state.
type DnfCheckUpdate struct{}

var updateFact = types.Fact{
	Name: "run.update_required",
	Tier: types.TierOne,
}

// Provides registers the fact key supplied by this source.
func (d *DnfCheckUpdate) Provides(p types.Provides) {
	p[updateFact.Name] = d
}

// Requires declares the external command needed.
func (d *DnfCheckUpdate) Requires(r *types.Requires) {
	r.Programs = append(r.Programs, "dnf check-update")
}

// Parse executes `dnf check-update` and records whether updates are required.
// It respects a config key `net` (bool). If present and false, network queries
// are disabled and a placeholder value is stored.
func (d *DnfCheckUpdate) Parse(f *types.FactDB) {
	if !global.Config.NetEnabled {
		updateFact.Value = types.NetDisabledMsg
		f.AddFact(updateFact)
		return
	}

	_, rc, _ := input.CommandRunner("dnf check-update")

	switch rc {
	case 0:
		updateFact.Value = "False"
	case 100:
		updateFact.Value = "True"
	default:
		updateFact.Value = types.ParseFailMsg
	}
	f.AddFact(updateFact)
}
